soroban-env-host/src/test/bls12_381.rs

```diff
@@ -0,0 +1,1277 @@
+        let vp1 = sample_host_vec::<G1Affine>(&host, G1_SERIALIZED_SIZE, 3)?;
+        let mut vp2 = sample_host_vec::<G2Affine>(&host, G2_SERIALIZED_SIZE, 3)?;
+        vp2 = host.vec_put(
+            vp2,
+            U32Val::from(1),
+            sample_g2_not_on_curve(&host)?.to_val(),
+        )?;
+        assert!(HostError::result_matches_err(
+            host.bls12_381_multi_pairing_check(vp1, vp2),
+            (ScErrorType::Crypto, ScErrorCode::InvalidInput)
+        ));
+    }
+    // 5. e(P, Q+R) = e(P, Q)*e(P, R)
+    {
+        host.budget_ref().reset_default()?;
+        let p = sample_g1(&host)?;
+        let neg_p = neg_g1(p, &host)?;
+        let q = sample_g2(&host)?;
+        let r = sample_g2(&host)?;
+        let q_plus_r = host.bls12_381_g2_add(q, r)?;
+
+        //check e(-P, Q+R)*e(P, Q)*e(P, R) == 1
+        let g1_vec = host.vec_new_from_slice(&[neg_p.to_val(), p.to_val(), p.to_val()])?;
+        let g2_vec = host.vec_new_from_slice(&[q_plus_r.to_val(), q.to_val(), r.to_val()])?;
+        let res = host.bls12_381_multi_pairing_check(g1_vec, g2_vec)?;
+        assert!(res.as_val().is_true())
+    }
+    // 6. e(P+S, R) = e(P, R)*e(S, R)
+    {
+        host.budget_ref().reset_default()?;
+        let p = sample_g1(&host)?;
+        let s = sample_g1(&host)?;
+        let r = sample_g2(&host)?;
+        let neg_r = neg_g2(r, &host)?;
+        let p_plus_s = host.bls12_381_g1_add(p, s)?;
+        // check e(P+S, -R) * e(P, R)*e(S, R) == 1
+        let g1_vec = host.vec_new_from_slice(&[p_plus_s.to_val(), p.to_val(), s.to_val()])?;
+        let g2_vec = host.vec_new_from_slice(&[neg_r.to_val(), r.to_val(), r.to_val()])?;
+        let res = host.bls12_381_multi_pairing_check(g1_vec, g2_vec)?;
+        assert!(res.as_val().is_true())
+    }
+
+    // 7. e([a]P, [b]Q) = e([b]P, [a]Q) = e([ab]P, Q)= e(P, [ab]Q)
+    {
+        host.budget_ref().reset_default()?;
+        let a = sample_fr(&host)?;
+        let b = sample_fr(&host)?;
+        let p = sample_g1(&host)?;
+        let neg_p = neg_g1(p, &host)?;
+        let q = sample_g2(&host)?;
+        let neg_q = neg_g2(q, &host)?;
+        let a_p = host.bls12_381_g1_mul(p, a)?;
+        let b_p = host.bls12_381_g1_mul(p, b)?;
+        let a_q = host.bls12_381_g2_mul(q, a)?;
+        let b_q = host.bls12_381_g2_mul(q, b)?;
+        let ab = host.bls12_381_fr_mul(a, b)?;
+        let ab_p = host.bls12_381_g1_mul(p, ab)?;
+        let ab_q = host.bls12_381_g2_mul(q, ab)?;
+        // check e([a]P, [b]Q) * e([b]P, [a]Q) * e([ab]P, -Q) * e(-P, [ab]Q) == 1
+        let g1_vec =
+            host.vec_new_from_slice(&[a_p.to_val(), b_p.to_val(), ab_p.to_val(), neg_p.to_val()])?;
+        let g2_vec =
+            host.vec_new_from_slice(&[b_q.to_val(), a_q.to_val(), neg_q.to_val(), ab_q.to_val()])?;
+        let res = host.bls12_381_multi_pairing_check(g1_vec, g2_vec)?;
+        assert!(res.as_val().is_true())
+    }
+    Ok(())
+}
+
+// fr arithmetics
+
+// serialization roundtrip
```

Comment:
More test vectors are probably fine as a followup, but I think serialization is probably among the most risky things here, so it would be nice to at least have some basic tests. Though I see you're testing bad input formats for a lot of functions (which is great!), so I wonder what's left.